src/main/java/org/alfresco/rest/api/impl/GroupsImpl.java

```diff
@@ -809,7 +810,15 @@ public void deleteGroupMembership(String groupId, String groupMemberId)
         }
 
         validateGroupMemberId(groupMemberId);
-        // TODO: Verify if groupMemberId is member of groupId
+
+        // Verify if groupMemberId is member of groupId
+        AuthorityType authorityType = AuthorityType.getAuthorityType(groupMemberId);
+        Set<String> containedAuthorities = authorityService.getContainedAuthorities(authorityType, groupId, true);

```

Comment:
Is this a scalable way of checking for membership? What happens if there are 10k people in the group? Is this the usual way we check for membership?

Reply:
It's true, it is not, unfortunately AuthorityService does not offer a method that we can use to do the check. I think we can return to the initial solution proposed in this comment https://issues.alfresco.com/jira/browse/REPO-1306?focusedCommentId=501391&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-501391.

Reply:
> “I am out at a conference - there is an in memory bridge structure used to answer most questions. If it uses that it will be fine” (Andy Hind)

@derekhulley -- do you have an opinion on whether we need to optimise this now or whether the current solution is ok, performance-wise?

Reply:
The implementation has to be efficient or we are simply creating serious problems for ourselves.  If there is no service that does the job, then one should be created.
I will have a look at the bridge table exposure.

Reply:
There is code that will give it to you but you'd have to use the DAO, similar to this in *AuthorityServiceImpl*:
    
    /**
     * Checks if the {@code authority} (normally a username) is the same as or is contained
     * within the {@code parentAuthority}.
     * @param authority String
     * @param parentAuthority a normalized, case sensitive authority name
     * @return {@code true} if does, {@code false} otherwise.
     */
    private boolean hasAuthority(String authority, String parentAuthority, Set<String> positiveHits, Set<String> negativeHits)
    {
        // Even users are matched case sensitively in ACLs
        if (AuthorityType.getAuthorityType(parentAuthority) == AuthorityType.USER)
        {
            return false;
        }

        if (parentAuthority.equals(authority))
        {
            return true;
        }

        return authorityDAO.isAuthorityContained(parentAuthority, authority, positiveHits, negativeHits);        
        
    }

If you make a change to the *5.2.N*, then exposing that on the service itself (without the Sets) would be OK.

Under the covers, the DAO uses a bridge table to look *upwards*, which is scalable.  Looking *down* the group hierarchy is not scalable, as Matt pointed out.

Reply:
Thank you both for the help. Since there is no way of doing this using already exposed methods, will create something like hasAuthority as you mentioned. It's also the one I was thinking of when writing this [comment](https://issues.alfresco.com/jira/browse/REPO-1306?focusedCommentId=501391&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-501391).